src/Microsoft.DotNet.Build.Tasks/NuGetPush.cs

```diff
@@ -0,0 +1,140 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using Microsoft.Build.Framework;
+using Microsoft.Build.Utilities;
+using System;
+using System.Collections.Generic;
+using System.Diagnostics;
+
+namespace Microsoft.DotNet.Build.Tasks
+{
+    /// <summary>
+    /// Run NuGet.exe to push a set of packages.
+    /// 
+    /// TimeoutPerPushInSeconds is a timeout per package push. This allows the task to retry when an
+    /// HTTP hang happens. A "NuGet.exe push *.nupkg -Timeout ###" call would fail without retrying.
+    /// https://github.com/NuGet/Home/issues/5224 requests retrying on hangs in NuGet.exe.
+    /// </summary>
+    public class NuGetPush : Task, ICancelableTask
+    {
+        [Required]
+        public string NuGetExePath { get; set; }
+
+        [Required]
+        public ITaskItem[] Inputs { get; set; }
+
+        [Required]
+        public string Source { get; set; }
+
+        public string ApiKey { get; set; }
+
+        public string Verbosity { get; set; } = "Detailed";
+
+        public int TimeoutPerPushInSeconds { get; set; } = TimeSpan.FromMinutes(5).Seconds;
+
+        public int RetriesPerPush { get; set; } = 5;
+
+        private bool _cancelRemainingPushes;
+
+        public void Cancel()
+        {
+            _cancelRemainingPushes = true;
+        }
+
+        public override bool Execute()
+        {
+            foreach (ITaskItem input in Inputs)
+            {
+                for (int i = 0; i < RetriesPerPush; i++)
+                {
+                    if (_cancelRemainingPushes)
+                    {
+                        break;
+                    }
+
+                    var args = new List<string>
+                    {
+                        "push", input.ItemSpec,
+                        "-Source", Source,
+                        "-Timeout", TimeoutPerPushInSeconds.ToString(),
+                        "-Verbosity", Verbosity
+                    };
+
+                    if (!string.IsNullOrEmpty(ApiKey))
+                    {
+                        args.Add("-ApiKey");
+                        args.Add(ApiKey);
+                    }
```

Comment:
Any reason these aren't outside of the retry loop?  It doesn't seem like the args need to be redefined on each try.